pkg/event/bus.go

```diff
@@ -69,6 +71,10 @@ func (bus *Bus) Trigger(evt string, evtErr error, context map[string]interface{}
 			}
 		}
 
+		if slices.Contains(hook.Events, "preapply") && evt == "presync" && hook.Executed {
```

Comment:
I think this special handling is unnecessary- when one included a hook for both events `preapply` and `presync`, almost probably it's what they wanted, for e.g. notifying the timing helmfile triggered respective events. There shouldn't be a hard-coded logic to block such usage.

Reply:
Or could you clarify the context behind this guard a bit more? 🙏

Reply:
@yxxhero @itscaro Do you have any insights on this? I'm wondering if I'm missing any practical use-cases of this feature that might make this special handling necessary

Reply:
I added it as a response to @gakesson, it felt reasonable, but I can remove it if not wanted.

The reasoning was that we can add both `preapply` and `presync` to make sure that the hook is run both if `apply` or `sync` is used, but that it only should run once. Since the `presync` hook is called when running apply as well, if there is any changes that will be synced to the cluster.

Reply:
@Sajfer @mumoshu it was mostly that to me it is not intuitive that if I run "helmfile apply" and have events as ["presync", "preapply"] that the hook would be executed twice (in case sync is invoked for the release), but perhaps this type of special handling is not consistent how events and hooks work today.
But I'm fine either way, as long as we have a well-defined and documented behavior.

Reply:
@gakesson Thanks for claryfing! The current behavior is that helmfile runs the hook on each event listed under the `events` field. If you listed both `presync` and `prepply`, helmfile will just run the hook for both. Does it make sense? Then we shall revert this special handling and better keep it consistent with all the other events.

Reply:
Also, that special handling would make the use-case I've explained in https://github.com/helmfile/helmfile/pull/79#discussion_r889647170 impossible

Reply:
Perhaps there's one more question that helps to distill it- when do you actually need to specify both `presync` and `preapply` for a single hook? I reread https://github.com/roboll/helmfile/issues/1940#issuecomment-1104785836 but still unsure.

`presync` can be used for anything that should be run before `helm upgrade --install`, and `preapply` can be used for anything that should be run on each release being processed by helmfile. In other words, releases targeted by `preapply` is a superset of `presync` targets. In case `presync` is insufficient, you'd only use `preapply`, not both? 🤔

Reply:
If we only have the `preapply` hook set and run `helmfile sync` then that hook would not run right? So I would not say it is a superset of `presync`. 
So if I want to have the hook running on all actions that affect the cluster, i.e `apply` and `sync` I would have it run twice in the case of `apply`, but only once if I run `sync`.

But if the consensus is that it is unnecessary I will remove this check.

Reply:
@mumoshu The use-case I was thinking of was that the hook execution would behave same (=execute once always) irrespective of our users use sync or apply (we deliver our helmfile to other parts of company org). In that case we would  specify ["preapply", "presync"] to cover both rollout commands, but obviously this can instead be solved by simply instructing users to only use apply and never sync if we use the preapply event. And at least for us it is not a problem to execute the hook twice, only a bit redundant.
But for consistency I think it is fine to not have this special handling. Sorry for the confusion on this discussion. :)

Reply:
> If we only have the preapply hook set and run helmfile sync then that hook would not run right? 

@Sajfer Ah, good point! And that's how it's implemented in this PR now, right? I'd say the new hook should be triggered also on `helmfile sync`.

It's unfortunately that presync and preapply doesn't correspond 1:1 to `helmfile sync` and `helmfile apply` respectively. But still, to better reflect how it's supposed to work:

> `preapply` hooks will trigger everytime a release is applied, indifferent if there is changes to the release or not.

we'd better make it triggered also on helmfile-sync.

Reply:
Correct, the current implementation only runs on `apply` and not on `sync`.

Just to be clear so there is no more changes after:
* we want the `preapply` hook to run both on `apply` and `sync`. 
* We do not want to have a prevention for the hook to run twice if both `preapply` and `presync` is added.

Is this correct?

Reply:
@Sajfer Yes, you're absolutely correct. I think the combination of those two changes makes the new preapplly hook more consistent with other hooks(Each hook is run only once per release on a lifecycle event of each release, regardless of which helmfile sub-command you're using). Does it make sense to you?